### PR TITLE
stacks: add ClusterStackRequests with ClusterRoles, change StackRequests to use Roles

### PR DESCRIPTION
**Pull request by displague**
Implementation of #672 

### Description of your changes

* [x] add ClusterStackRequests with ClusterRoles
* [x] change StackRequests to StackInstall (Same for ClusterStackRequests)
  * [x] after rebasing, reapply these changes
* [x] add permissionScope field (accepts "Cluster" or "Namespaced")
* [x] enforce Namespace scoped Stack CRDs
* [x] generate RBAC rules from owned and dependent CRDs. Include some common rules.
* [x] add tests for ClusterStackInstall
* [x] reduce duplication of code between StackInstall and ClusterStackInstall
* [x] ClusterStackInstalls produce ClusterRoles/Bindings, StackInstalls produce Roles/Bindings 
* [x] fixes the Ready field in Stack and StackInstall print columns

### After Deploy

Verify and proceed from draft:

* crossplaneio/stack-gcp#19
* crossplaneio/stack-azure#5
* crossplaneio/stack-aws#5
* crossplaneio/sample-stack#2
* crossplaneio/sample-stack-wordpress#8
* crossplaneio/crossplane-cli#11

We may want to merge https://github.com/crossplaneio/crossplane/pull/772 immediately after this PR and update all of those PRs with both sets of Stacks changes.

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [x] Updated any relevant [documentation], [examples], or [release notes].
- [x] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplaneio/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

### CODE REVIEW

File: apis/stacks/v1alpha1/types.go [hunk 1]
```diff
@@ -82,6 +82,57 @@ type StackRequestStatus struct {
 
 // +kubebuilder:object:root=true
 
+// ClusterStackRequest is the CRD type for a request to add a stack to Crossplane.
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".status.conditions[?(@.type==Ready)].status"
```
Comment (soorena776) [inline apis/stacks/v1alpha1/types.go]: I think you would need to wrap the *Ready* with single quotes:
`// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".status.conditions[?(@.type=='Ready')].status"`

File: pkg/controller/stacks/clusterrequest/clusterrequest.go [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+}
+
+// Controller is responsible for adding the ClusterStackRequest
+// controller and its corresponding reconciler to the manager with any runtime configuration.
+type Controller struct{}
+
+// SetupWithManager creates a new Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (c *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	r := &Reconciler{
```
Comment (soorena776) [inline pkg/controller/stacks/clusterrequest/clusterrequest.go]: Should we use the managed reconciler pattern?
https://github.com/crossplaneio/crossplane/blob/master/pkg/controller/aws/cache/managed.go#L64

File: pkg/controller/stacks/install/stackinstall.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+// Reconcile reads that state of the StackInstall for a Instance object and makes changes based on the state read
+// and what is in the Instance.Spec
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("reconciling", "kind", r.kubeobject.GroupVersionKind(), "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	// fetch the CRD instance
+	if err := r.kube.Get(ctx, req.NamespacedName, r.kubeobject); err != nil {
```
[the remark this answers is not in the document]
Comment (displague) [reply pkg/controller/stacks/install/stackinstall.go]: I've moved the creation of the type (and getting of that reference) to within the Reconcile function.  My last few attempts only inched the instantiation closer. :mouse2: :mountain:

**Comment by jbw976** [comment]
great work @displague, everything looks good now to me.  Once you've confirmed that our wild card parsing of dependsOn CRDs works OK, please let me know and I'll mark this approved.

If the wild card parsing doesn't work, that would probably be OK for v0.3 actually.  We would just not include any documentation/examples that uses wild cards, and we would not put wild cards into any of the stacks that we are building.  Let me know what you find!

**Comment by displague** [comment]
@jbw976 Wildcards fit in easily. https://github.com/crossplaneio/crossplane/pull/729/commits/d59d2c646a6e39fd2fb43ee805048f9649c8b245

**Pull request merged**